### PR TITLE
uds: chage Mutex to RWMutex for fast already-connected path (rebase)

### PR DESCRIPTION
**Pull request by truthbk**
This PR rebases #74 and includes the original changes submitted by @KJTsanaktsidis, it's merely a rebase after the introduction of the async non-blocking writer with #81

Thank you for your contribution @KJTsanaktsidis

### CODE REVIEW

File: statsd/uds_blocking.go [hunk 1]
```diff
@@ -63,3 +59,34 @@ func (w *blockingUdsWriter) Close() error {
 	}
 	return nil
 }
+
+func (w *udsWriter) ensureConnection() (net.Conn, error) {
```
Comment (arbll) [inline statsd/uds_blocking.go]:
```suggestion
func (w * blockingUdsWriter) ensureConnection() (net.Conn, error) {
```

File: statsd/uds_blocking.go [hunk 2]
```diff
@@ -63,3 +59,34 @@ func (w *blockingUdsWriter) Close() error {
...
+
+	newConn, err := net.Dial(w.addr.Network(), w.addr.String())
+	if err != nil {
+		return nil, err
+	}
+	w.conn = newConn
+	return newConn, nil
+}
+
+func (w *udsWriter) unsetConnection() {
```
Comment (arbll) [inline statsd/uds_blocking.go]:
```suggestion
func (w * blockingUdsWriter) unsetConnection() {
```

**Comment by KJTsanaktsidis** [comment]
Nice - thanks for merging!

Quick question, with the inclusion of the non-blocking version, is there any compelling reason to use the blocking one with this mutex over the non-blocking one other than error handling?

If not, I'll probably switch our app over to use the non-blocking one

**Comment by arbll** [comment]
Hi @KJTsanaktsidis 

No reason apart from error checking. We are keeping it as the default implementation in 2.x and plan to release a new major that will make the async version the default and only implementation.

**Pull request merged**